### DISCUSSION TITLE
Build RPM with AppStream data - not quite working in Software Center - help?

### DISCUSSION THREAD

**Discussion by kuiperzone** [ximion/appstream#468] [Q&A]
Hi,

I've exhausted all options and I need to ask a question. I am actually trying to build a binary RPM with a test Hello World app (called "HelloPuppy"), but I suspect this is an APPSTREAM metainfo problem.

I want my app to show in the Gnome Software Center, which is why I am including a metainfo.xml file.

Now, I have managed to do almost everything all by my little self. I can install and remove the app using "dnf install/remove", and when install it does indeed show in the Software Center (see screenshot).

BUT AND HOWEVER....

Clicking on "Open" or "Uninstall" in the apps' page in the Software Center does nothing. In other wolds, the page in the Software the Center does not seem to link properly to the application (although it shows the name, icon etc.). I cannot figure out what I am missing or doing wrong.

SEE SCREENSHOTS. Shows Software Center, and location of .desktop and metainfo.xml files.

![Screenshot 1](https://user-images.githubusercontent.com/17156238/222931232-46bc46c3-480c-467d-9a99-9f7a138a3a71.png)
![Screenshot 2](https://user-images.githubusercontent.com/17156238/222931233-134ff48e-e329-4383-9b2b-18f2ee9d6bdf.png)

Here is my desktop file: /usr/share/applications/net.example.hellopuppy.desktop

  [Desktop Entry]
  Type=Application
  Name=Hello Puppy
  Icon=net.example.hellopuppy
  Comment=A HelloWorld application
  Exec=/opt/net.example.hellopuppy/HelloPuppy
  TryExec=/opt/net.example.hellopuppy/HelloPuppy
  Terminal=true
  Categories=Utility
  MimeType=
  Keywords=

Here is my metadata.xml file: /usr/share/metainfo/net.example.hellopuppy.metainfo.xml

    <?xml version="1.0" encoding="UTF-8"?>
    <component type="desktop">
          <id>net.example.hellopuppy</id>
        <name>HelloPuppy</name>
        <summary>A HelloWorld application</summary>
        <developer_name>Acme Ltd</developer_name>
        <url type="homepage">https://example.net</url>
        <metadata_license>MIT</metadata_license>
        <project_license>GPL-3.0-or-later</project_license>
    
        <content_rating type="oars-1.1" />
    
        <description>
            <p>This is a longer application description. REPLACE THIS WITH YOUR OWN.
            IMPORTANT: In this file, you can use supported macros. See the --help output for details.</p>
        </description>
    
        <releases>
            <release version="1.0.0" date="2023-03-04">
                <description><p>The latest release.</p></description>
            </release>
        </releases>
    
        <screenshots>
            <screenshot type="default">
                <image>https://i.postimg.cc/0jc8xxxC/Hello-Computer.png</image>
            </screenshot>
        </screenshots>
    
        <launchable type="desktop-id">net.example.hellopuppy.desktop</launchable>
    
    </component>

**Comment by ximion** [comment]
I don't think this is an AppStream problem - your metadata looks perfectly fine and even validates cleanly with `appstreamcli validate`. The only hint I'd have is to replace `type="desktop"` with `type="desktop-application"` (the latter is newer), but changing that does not change the behavior whatsoever.

So, this is most certainly an issue with either GNOME Software or Fedora (or RHEL/CentOS or whatever OS you are using that uses dnf). Where did you install your icon to? (I noticed your app doesn't have an icon), and which version of Fedora and GNOME Software are you using? Potentially you'll need to file a bug with either of them.
GNOME Software's verbose log is likely also very helpful (`killall gnome-software ; gnome-software --verbose`)

**Reply by kuiperzone** [reply]
See below

**Comment by kuiperzone** [comment]
Thank you for the reply.

I'm running Fedora 37.

I'm not convinced this is a Gnome Software issue (though it might be). I've tested the .rpm file for the KeepassXC popular app, for example. I've installed and removed it using dnf and the KeepassXC app works fine in Gnome Software. I've extracted its desktop and metainfo files and compared them with my own. I cannot see what I am doing wrong however! How strange!

Bizarrely in my metainfo (above) file, I defined the Name of my app to be "HelloPuppy" (no space), but Gnome Software shows it as "Hello Puppy" with space. Presumably this is coming from the desktop file.

The only thing I'm doing differently perhaps is that I'm installing my main application and all related files in its own directory under /opt/appname directory. I'm doing this because my app is a dotnet C# app which comes with a load of DLLs and associated runtime files. I am, however, put a small bash script file in /usr/bin to launch the app.  So my RPM (below) is simply packing the output of the dotnet build.

My meta info is here, including icons:

/usr/share/applications/net.example.hellopuppy.desktop
/usr/share/metainfo/net.example.hellopuppy.metainfo.xml
/usr/share/icons/hicolor/48x48/apps/net.example.hellopuppy.png
/usr/share/icons/hicolor/32x32/apps/net.example.hellopuppy.png
/usr/share/icons/hicolor/24x24/apps/net.example.hellopuppy.png
/usr/share/icons/hicolor/16x16/apps/net.example.hellopuppy.png
/usr/share/icons/hicolor/scalable/apps/net.example.hellopuppy.svg


My Spec file:

    Name: HelloPuppy
    Version: 1.0.0
    Release: 1
    BuildArch: x86_64
    Summary: A HelloWorld application
    License: GPL-3.0-or-later
    Vendor: Acme Ltd
    Url: https://example.net
    
    AutoReqProv: no
    BuildRequires: libappstream-glib
    
    %check
    desktop-file-validate %{buildroot}/%{_datadir}/applications/*.desktop
    appstream-util validate-relax --nonet %{buildroot}%{_metainfodir}/net.example.hellopuppy.metainfo.xml
    
    %description
    A HelloWorld application
    
    %files
    /opt/net.example.hellopuppy/Microsoft.CSharp.dll
    /opt/net.example.hellopuppy/HelloPuppy.dll
    /opt/net.example.hellopuppy/HelloPuppy.deps.json
    
    [REMOVED MOST APP FILES TO SHORTEN POST]
    
    /opt/net.example.hellopuppy/libmscordaccore.so
    /opt/net.example.hellopuppy/libmscordbi.so
    /opt/net.example.hellopuppy/HelloPuppy
    /usr/bin/hellopuppy
    /usr/share/applications/net.example.hellopuppy.desktop
    /usr/share/metainfo/net.example.hellopuppy.metainfo.xml
    /usr/share/icons/hicolor/48x48/apps/net.example.hellopuppy.png
    /usr/share/icons/hicolor/32x32/apps/net.example.hellopuppy.png
    /usr/share/icons/hicolor/24x24/apps/net.example.hellopuppy.png
    /usr/share/icons/hicolor/16x16/apps/net.example.hellopuppy.png
    /usr/share/icons/hicolor/scalable/apps/net.example.hellopuppy.svg

**Reply by kuiperzone** [reply]
Here is my test RPM (inside a zip file):

dnf install / remove work fine.

[HelloPuppy-1.0.0-1.x86_64.zip](https://github.com/ximion/appstream/files/10893276/HelloPuppy-1.0.0-1.x86_64.zip)

The app shows in the desktop (and arc menu). After re-starting gnome-software, it shows the installed page (shown above). But the buttons I refer to don't work. dnf remove to remove it.

**Comment by kuiperzone** [comment]
And to be clear -- _if I install an app from the command line using `dnf install packname.rpm`, will it actually be removable from Gnome Software_? Or is this just normal behaviour (perhaps the buttons should be greyed out in this case).

For example, is the following line in desktop file relevant or necessary:

X-AppInstall-Package=gnome-software

I see the above in some packages?

**Comment by ximion** [comment]
All your metadata looks fine, I think you are really hitting a GNOME Software bug here. Please have a look at the verbose output of GS, that might give you a clue!

However:
> Bizarrely in my metainfo (above) file, I defined the Name of my app to be "HelloPuppy" (no space), but Gnome Software shows it as "Hello Puppy" with space. Presumably this is coming from the desktop file.

This is extremely odd! Have a look at your system and check if there isn't some old metadata somewhere else that interferes, for example in your user directory (`~/.local` & co.) or in `/usr/local/share`.

> For example, is the following line in desktop file relevant or necessary:
> X-AppInstall-Package=gnome-software

No, this was used by an old Ubuntu feature and precursor of AppStream and has not been in use for almost a decade. So you can just ignore this.

> And to be clear -- if I install an app from the command line using dnf install packname.rpm, will it actually be removable from Gnome Software?

Yes - you should be able to remove it via GS, no matter how it was installed.

**Reply by kuiperzone** [reply]
>killall gnome-software ; gnome-software --verbose

Yes this was helpful! Here are some results:

Clicking Open:

00:03:50:768 Gs  running launch on plugin=packagekit with dedupe-flags=7 with interactive=True on apps system/*/*/net.example.hellopuppy/*

Not sure this is looking for.

My actuall runnable binary is defined by the destkop line:

Exec=/opt/net.example.hellopuppy/HelloPuppy


Clicking Uninstall:

00:04:40:678 Gs  uninstall net.example.hellopuppy
00:04:40:678 Gs  Chaining cancellation from 0x5651f024c560 to 0x7f8438000ea0
00:04:40:678 Gs  nothing adopted system/*/*/net.example.hellopuppy/*
00:04:40:678 Gs  Setting I/O priority of thread 0x7f84742eee40 to IDLE, 7
00:04:40:679 Gs  Setting I/O priority of thread 0x5651ed1da360 to IDLE, 7
00:04:40:680 Gs  Setting I/O priority of thread 0x7f8478002300 to IDLE, 7
00:04:40:680 GsPluginPackageKit ignoring /usr/share/applications/net.example.hellopuppy as does not exist
00:04:40:680 Gs  Setting I/O priority of thread 0x5651ed1d5860 to IDLE, 7
00:04:40:680 Gs  running unknown with dedupe-flags=7, elapsed time since creation 2ms
00:04:40:680 Gs  running remove on plugin=packagekit with dedupe-flags=7 with refine-flags=require-setup-action,require-origin with interactive=True on apps system/*/*/net.example.hellopuppy/*, elapsed time since creation 2ms
00:04:40:692 Gs  content rating system is guessed as PEGI from en_GB.UTF-8
00:04:40:692 Gs  content rating system is guessed as PEGI from en_GB.UTF-8

Am I correct in thinking it is looking for a packaged called: net.example.hellopuppy ?

My RPM was named: HelloPuppy

and I use: `sudo rpm remove HelloPuppy` to remove it. 

Is this the problem?

**Comment by kuiperzone** [comment]
HI @ximion , not heard back.

Despite  debug from Gs, still not clear why this is failing.

00:03:50:768 Gs running launch on plugin=packagekit with dedupe-flags=7 with interactive=True on apps system///net.example.hellopuppy/*

I don't know what this means?

Likewise, I do not know why it seems to be trying to uninstall a package called "net.example.hellopuppy" (tha app ID) when the RPM package is called "HelloPuppy"?

Any ideas?
